sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleHistogramAggregator.java

```diff
@@ -0,0 +1,168 @@
+    }
+
+    long[] mergedCounts = new long[x.getCounts().size()];
+    for (int i = 0; i < x.getCounts().size(); ++i) {
+      mergedCounts[i] = x.getCounts().get(i) + y.getCounts().get(i);
+    }
+    return HistogramAccumulation.create(
+        x.getCount() + y.getCount(),
+        x.getSum() + y.getSum(),
+        x.getBoundaries(),
+        Arrays.stream(mergedCounts).boxed().collect(Collectors.toList()));
+  }
+
+  @Override
+  public final MetricData toMetricData(
+      Map<Labels, HistogramAccumulation> accumulationByLabels,
+      long startEpochNanos,
+      long lastCollectionEpoch,
+      long epochNanos) {
+    return MetricData.createDoubleHistogram(
+        getResource(),
+        getInstrumentationLibraryInfo(),
+        getInstrumentDescriptor().getName(),
+        getInstrumentDescriptor().getDescription(),
+        getInstrumentDescriptor().getUnit(),
+        DoubleHistogramData.create(
+            isStateful() ? AggregationTemporality.CUMULATIVE : AggregationTemporality.DELTA,
+            MetricDataUtils.toDoubleHistogramPointList(
+                accumulationByLabels,
+                isStateful() ? startEpochNanos : lastCollectionEpoch,
+                epochNanos)));
+  }
+
+  @Override
+  public HistogramAccumulation accumulateDouble(double value) {
+    return HistogramAccumulation.create(
+        1, value, Collections.emptyList(), Collections.singletonList(1L));
+  }
+
+  @Override
+  public HistogramAccumulation accumulateLong(long value) {
+    return HistogramAccumulation.create(
+        1, value, Collections.emptyList(), Collections.singletonList(1L));
+  }
+
+  static final class Handle extends AggregatorHandle<HistogramAccumulation> {
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    @GuardedBy("lock")
+    private final State current;
+
+    Handle(double[] boundaries) {
+      current = new State(boundaries);
+    }
+
+    @Override
+    protected HistogramAccumulation doAccumulateThenReset() {
+      lock.writeLock().lock();
+      try {
+        HistogramAccumulation result =
+            HistogramAccumulation.create(
+                current.count,
+                current.sum,
+                Arrays.stream(current.boundaries).boxed().collect(Collectors.toList()),
+                Arrays.stream(current.counts).boxed().collect(Collectors.toList()));
+        current.reset();
+        return result;
+      } finally {
+        lock.writeLock().unlock();
+      }
+    }
+
+    @Override
+    protected void doRecordDouble(double value) {
+      int bucketIndex = current.findBucketIndex(value);
+
+      lock.writeLock().lock();
+      try {
+        current.record(bucketIndex, value);
+      } finally {
+        lock.writeLock().unlock();
+      }
+    }
+
+    @Override
+    protected void doRecordLong(long value) {
+      doRecordDouble((double) value);
+    }
+
+    private static final class State {
+      private long count;
+      private double sum;
```

Comment:
since these two values can always be derived from the contents of the buckets, could we not keep them in the live state, and only calculate them when we need to report/merge the data, which happens off the hot path?

Reply:
if we did that, we could only lock the specific bucket, rather than having to lock the whole data structure, when we want to record a value.

Reply:
nice one. will do.